src/coreclr/pal/src/include/pal/virtual.h

```diff
@@ -190,7 +191,13 @@ class ExecutableMemoryAllocator
     // This constant represent the max size of the virtual memory that this allocator
     // will try to reserve during initialization. We want all JIT-ed code and the
     // entire libcoreclr to be located in a 2GB range.
-    static const int32_t MaxExecutableMemorySize = 0x7FFF0000;
+    static const int32_t MaxExecutableMemorySize = 0x7FFF0000;// 2GB - 64KB
+#else
+    // Smaller values for ARM64 where relative calls/jumps only work within 128MB
+    static const int32_t CoreClrLibrarySize = 32 * 1024 * 1024;
+    static const int32_t MaxExecutableMemorySize = 0x7FF0000; // 128MB - 64KB
```

Comment:
How much of this range gets consumed in a real-world ASP.NET app? 

Note that we map all sorts of stuff into this range, including R2R images. I would expect that 128MB gets exhausted fairly quickly given how things work today.

I agree that this fix works well for micro-benchmarks that are very unlikely to exhaust the 128MB range.

Reply:
Will try to investigate, I guess the idea that the hot code (tier1) should better be closer to VM's methods it uses.
For large apps I guess we still want to try to address your suggestion with emitting direct addresses without jump-stubs in tier1 https://github.com/dotnet/runtime/issues/62302

For R2R-only it can be improved by pgo + method sorting?

Reply:
I do not think R2R code today benefits from being close to coreclr as all 'external' calls/branches have to go through indirection cells anyway. This may have been different in the days of fragile ngen?
Please correct me if I'm wrong @jkotas.

Reply:
> For large apps I guess we still want to try to address your suggestion with emitting direct addresses without jump-stubs

I do not think it is just for large apps. With TC enabled, all managed->managed method calls go through a precode that has the exact same instructions as jump stub and so it will introduce similar bottleneck as what you have identified.

> For R2R-only it can be improved by pgo + method sorting?

R2R images are generally smaller than 128MB. You can only sort within the image, so the sorting won't help with jump stubs. (Sorting within image is still good for locality.)

Also, once we get this all fixed, we may want to look at retuning the inliner. My feeling is that the inliner expands the code too much these days. Some of it may be just compensating for the extra method call overhead that we are paying today.

Reply:
> I do not think R2R code today benefits from being close to coreclr as all 'external' calls/branches have to go through indirection cells anyway.

Calls from runtime generated stubs and JITed code to R2R code still benefit from the two being close.

Reply:
> JITed code to R2R code

Do these not go through an indirection when tiering is enabled?

Reply:
> Do these not go through an indirection when tiering is enabled?

Yes - when tiering is enabled. No - when tiering is disabled.

Reply:
> With TC enabled, all managed->managed method calls go through a precode that has the exact same instructions as jump stub and so it will introduce similar bottleneck as what you have identified.

I'll start from your suggestion to emit direct calls for T1 Caller calls T1 Callee (not as part of this PR)